packages/SwingSet/src/kernel/virtualObjectManager.js

```diff
@@ -0,0 +1,328 @@
+        return undefined;
+      } else {
+        const { type, virtual } = parseVatSlot(instanceKey);
+        if (type === 'object' && virtual) {
+          return `ws${storeID}.${instanceKey}`;
+        } else {
+          return undefined;
+        }
+      }
+    }
+
+    return harden({
+      has(key) {
+        const vkey = virtualObjectKey(key);
+        if (vkey) {
+          return !!syscall.vatstoreGet(vkey);
+        } else {
+          return backingMap.has(key);
+        }
+      },
+      init(key, value) {
+        const vkey = virtualObjectKey(key);
+        if (vkey) {
+          assert(
+            !syscall.vatstoreGet(vkey),
+            details`${q(keyName)} already registered: ${key}`,
+          );
+          syscall.vatstoreSet(vkey, JSON.stringify(m.serialize(value)));
+        } else {
+          assertKeyDoesNotExist(key);
+          backingMap.set(key, value);
+        }
+      },
+      get(key) {
+        const vkey = virtualObjectKey(key);
+        if (vkey) {
+          const rawValue = syscall.vatstoreGet(vkey);
+          assert(rawValue, details`${q(keyName)} not found: ${key}`);
+          return m.unserialize(JSON.parse(rawValue));
+        } else {
+          assertKeyExists(key);
+          return backingMap.get(key);
+        }
+      },
+      set(key, value) {
+        const vkey = virtualObjectKey(key);
+        if (vkey) {
+          assert(
+            syscall.vatstoreGet(vkey),
+            details`${q(keyName)} not found: ${key}`,
+          );
+          syscall.vatstoreSet(vkey, JSON.stringify(m.serialize(harden(value))));
+        } else {
+          assertKeyExists(vkey);
+          backingMap.set(key, value);
+        }
+      },
+    });
+  }
+
+  /**
+   * Make a new kind of virtual object.
+   *
+   * @param {*} instanceMaker  A function of the form `instanceMaker(state)` that
+   *    will return a representative instance wrapped around the given state.
+   *
+   * @returns {*} a maker function that can be called to manufacture new
+   *    instance of this kind of object.  The parameters of the maker function
+   *    are those of the `initialize` method implemented in the representative
+   *    produced by the `instanceMaker` parameter.
+   */
+  function makeKind(instanceMaker) {
+    const kindID = `${allocateExportID()}`;
+    let nextInstanceID = 1;
+
+    function makeRepresentative(innerSelf) {
+      function ensureState() {
+        if (!innerSelf.rawData) {
+          innerSelf = cache.lookup(innerSelf.instanceKey);
+        }
+      }
+
+      function wrapData() {
+        const activeData = {};
+        for (const prop of Object.getOwnPropertyNames(innerSelf.rawData)) {
+          Object.defineProperty(activeData, prop, {
+            get: () => {
+              ensureState();
+              return m.unserialize(innerSelf.rawData[prop]);
+            },
+            set: value => {
+              ensureState();
+              innerSelf.rawData[prop] = m.serialize(value);
```

Comment:
The degree of paranoia in that analysis is both impressive and disturbing.  I suspect it's not an issue because `serialize` chokes on accessors, but it's an easy change and better safe than sorry.  Done.

Reply:
Ah, good point, I didn't realize `marshal` refused that. I'll add a test to packages/marshal to make that behavior more visible.

I suspect a `Proxy` might be allowed.. I'll check that in `marshal` too.

Reply:
`marshal` cannot reliably detect proxies, and does not attempt to. The same TOCTOU issue remains. We still need to serialize up front.

Reply:
Yup, my test passes.. `marshal` cannot distinguish a `Proxy` from a regular object, so the caller can use one to get control during serialization.